.github/workflows/test.yml

```
@@ -77,6 +80,35 @@ jobs:
           files: service-${{ matrix.service }}/build/*.xml
           test_file_prefix: +service-${{ matrix.service }}/
           check_name: ${{ matrix.service }} ${{ matrix.suite }} tests
+      - name: archive test results
+        uses: actions/upload-artifact@v4.6.0
+        with:
+          name: service-${{ matrix.service }}
+          path: service-${{ matrix.service }}/build
+        if: matrix.suite == 'unit-test'
```

This may be something you tried and there's a reason it doesn't work, but I wonder if you could do the CodeCov upload here to avoid needing a save/download sequence.
```suggestion
      - name: service ${{ matrix.service }} codecov upload
        uses: codecov/codecov-action@v5
        if: matrix.suite == 'unit-test'
        with:
          name: paper-id-${{ matrix.service }}
          directory: ./service-${{ matrix.service }}
          fail_ci_if_error: false
          flags: paper-id-${{ matrix.service }}
          verbose: true
          token: ${{ secrets.CODECOV_TOKEN }}
```

---

ah this is nicer, have updated and seems to be working fine